### PR TITLE
impl Not for !

### PR DESCRIPTION
**Pull request by dtolnay**
The lack of this impl caused trouble for me in some degenerate cases of macro-generated code of the form `if !$cond {...}`, even without `feature(never_type)` on a stable compiler. Namely if `$cond` contains a `return` or `break` or similar diverging expression, which would otherwise be perfectly legal in boolean position, the code previously failed to compile with:

```console
error[E0600]: cannot apply unary operator `!` to type `!`
   --> library/core/tests/ops.rs:239:8
    |
239 |     if !return () {}
    |        ^^^^^^^^^^ cannot apply unary operator `!`
```

### CODE REVIEW

**Comment by dtolnay** [comment]
It looks like highfive hasn't assigned this.
r? rust-lang/libs-api

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by joshtriplett** [comment]
Nominating for the lang team because the libs-api team would like to check whether lang considers it a lang matter if we add impls of traits for `!`.

**Comment by scottmcm** [comment]
Is this the first trait impl on `!` for which there's an associated type?

`Output = !` does seem reasonable for `Not`, but I wonder if this would be a good time to think about how to pick it more generally.  Like `Iterator for !` is particularly non-obvious to me.

So I don't really have any particular concern here, but this seems like it's setting a precedent for other things -- does `Add for !` come next? -- where I'm not sure exactly what form it should have.

If there was going to be a lang thing here, it could be that the type checker just was smart about `!` specifically.  Since we might want something like `impl<T> Add<!> for T` and `impl<T> Add<T> for !`, but we can't have both of those.  Maybe the answer there is that we want features for that overlap anyway?

**Comment by Amanieu** [comment]
Regarding `Add`, can we just have a single `impl Add<!> for !` and let type coercion/inference do the rest?

File: library/core/src/ops/bit.rs [hunk 1]
```diff
@@ -68,6 +68,17 @@ macro_rules! not_impl {
 
 not_impl! { bool usize u8 u16 u32 u64 u128 isize i8 i16 i32 i64 i128 }
 
+#[stable(feature = "not_never", since = "1.58.0")]
```
Comment (teor2345) [inline library/core/src/ops/bit.rs:71]: It looks like this PR missed 1.58.
```suggestion
#[stable(feature = "not_never", since = "1.59.0")]
```

**Pull request merged**